### ISSUE TITLE
Empty inventory returned when having one or more machines with internal IPs only

### ISSUE THREAD

**Issue by Wnthr** [ansible-collections/hetzner.hcloud#178]
##### SUMMARY
I have machines running only with internal IPs for some of my infrastructure. When trying to use the inventory module, it fails with an error.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
hetzner.hcloud.hcloud inventory

##### ANSIBLE VERSION

```
% ~/venvs/ansible/bin/ansible --version
ansible [core 2.14.1]
  config file = /home/mw/src/ansible-hetzner-bootstrap/ansible.cfg
  configured module search path = ['/home/mw/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/mw/venvs/ansible/lib/python3.10/site-packages/ansible
  ansible collection location = /home/mw/.ansible/collections:/usr/share/ansible/collections
  executable location = /home/mw/venvs/ansible/bin/ansible
  python version = 3.10.8 (main, Nov 27 2022, 18:25:57) [GCC 11.3.0] (/home/mw/venvs/ansible/bin/python)
  jinja version = 3.1.2
  libyaml = True
```

##### COLLECTION VERSION

```
hetzner.hcloud                1.9.0
```

##### CONFIGURATION

```
CONFIG_FILE() = /home/mw/src/ansible-hetzner-bootstrap/ansible.cfg
DEFAULT_HOST_LIST(/home/mw/src/ansible-hetzner-bootstrap/ansible.cfg) = ['/home/mw/src/ansible-hetzner-bootstrap/inventory']
```

##### OS / ENVIRONMENT
inventory/hcloud.yaml:
`plugin: hcloud`

##### STEPS TO REPRODUCE
1. Have a machine without an external IP (neither ip4 nor ip6) in your hetzner cloud account
2. `% ansible-inventory -i inventory/hcloud.yaml --list`

##### EXPECTED RESULTS
An inventory output json populated with the machines in the cloud account, showing the internal IPs in that structure.

##### ACTUAL RESULTS
 ```
[WARNING]:  * Failed to parse /home/mw/src/ansible-hetzner-bootstrap/inventory/hcloud.yaml with auto plugin: 'NoneType' object has no attribute 'ip'
[WARNING]:  * Failed to parse /home/mw/src/ansible-hetzner-bootstrap/inventory/hcloud.yaml with yaml plugin: Plugin configuration YAML file, not YAML inventory
[WARNING]:  * Failed to parse /home/mw/src/ansible-hetzner-bootstrap/inventory/hcloud.yaml with ini plugin: Invalid host pattern 'plugin:' supplied, ending in ':' is not allowed, this character is reserved to provide a port.
[WARNING]: Unable to parse /home/mw/src/ansible-hetzner-bootstrap/inventory/hcloud.yaml as an inventory source
[WARNING]: No inventory was parsed, only implicit localhost is available
{
    "_meta": {
        "hostvars": {}
    },
    "all": {
        "children": [
            "ungrouped"
        ]
    }
}
```

**Comment by Wnthr** [comment]
It seems that is a reasonable default to me, but in this case, I am running the whole thing from inside an airgapped net, and that default doesn't work for me, but neither does using a hostname attribute, since I am not having dns records for that machine. It doesn't seem that your proposed solution is actually solving the problem of the plugin being usable in my situation?

Either way, it seems like regardless of whether you think the plugin should be able to solve my problem, it still breaks the plugin for both these machines and for my normal machines with an external IP, since the code breaks if any machine in the project doesn't have an external IP address, which means that I couldn't use the plugin to handle those machines either as it stands.

I wrote a quick small script that basically does what I need and grabs the first private_net and takes the IP from that if there is no external ip available. It seems to me that putting in some logic like that would make the plugin usable in both the cases described above. If that is not palatable, at least handling the fact that there are no primary ip by putting in the empty string seems to be a more robust way of handling a situation which would affect anyone who has at least one machine without a primary ipv4 address for their machine?

**Comment by maestroi** [comment]
I have the same problem even with nodes with an public and private ip.

tried different options but always same result



hetzner.hcloud                1.9.0
ansible [core 2.12.9]

`[WARNING]: Ansible is being run in a world writable directory (PATH), ignoring it as an ansible.cfg source. For more
information see https://docs.ansible.com/ansible/devel/reference_appendices/config.html#cfg-in-world-writable-dir
[WARNING]:  * Failed to parse PATH/hcloud.yml with script plugin: problem running
PATH/hcloud.yml --list ([Errno 8] Exec format error: 'PATH/hcloud.yml')
[WARNING]:  * Failed to parse /hcloud.yml with auto plugin: 'NoneType' object has no attribute 'items'
[WARNING]:  * Failed to parse PATH/hcloud.yml with yaml plugin: Plugin configuration YAML file, not YAML inventory
[WARNING]:  * Failed to parse PATH/hcloud.yml with ini plugin: Invalid host pattern 'plugin:' supplied, ending in ':' is not
allowed, this character is reserved to provide a port.
[WARNING]: Unable to parse PATH/hcloud.yml as an inventory source
[WARNING]: No inventory was parsed, only implicit localhost is available
{
    "_meta": {
        "hostvars": {}
    },
    "all": {
        "children": [
            "ungrouped"
        ]
    }
}`

hcloud.yml
`
plugin: "hcloud"
token: "XXX"
`

**Comment by apricote** [comment]
@maestroi It looks like you are encountering a different issue:

```
[WARNING]:  * Failed to parse /hcloud.yml with auto plugin: 'NoneType' object has no attribute 'items'
```

I am going to move this into a seperate issue.

**Comment by maestroi** [comment]
@apricote yeah sorry seemed related Sorry!

**Comment by apricote** [comment]
> - Only set the ansible_host to what connect_with specifies if the value is actually available, possible log warning otherwise
> - Fix the compose-ability to allow users to decide how they want to connect to instances using jinja2 templates

This works partially.

You can set the following:

```
plugin: hcloud
strict: true

network: 2395818
connect_with: hostname

compose:
  ansible_host: ipv4 | default(private_ipv4)
```

And the hosts will use the public ipv4 address if available, and the `private_ipv4` if not. Only issue is, that the argument `network` needs to be set to populate the `private_ipv4` field, but this also filters the list of server to only servers that are attached to this network.

We probably should return all networks somewhere in the host variables, so users can choose which one they want in compose/jinja2

**Comment by Wnthr** [comment]
> @Wnthr I am not an ansible expert, so please bear with me for the following questions:

Not at all, it is my pleasure to clarify what I am looking for!

> What do you do with the returned hosts from the inventory? Do you connect to these instance directly by using the inventory?

Yes, I connect with them directly when accessing the api from the inside via a proxy. When I connect from the outside, I use a bastion host to connect, which is configured in the ssh settings used on the machine using the inventory.

> What do you think about using multiple inventories for your public and private instances?

Well, my main contention is that if there exists a host without an external IP in the list, the inventory breaks. I don't mind having separate inventories in that I don't really see a workflow where I would work with both machines with and without an external address, but it's not clear to me how to make that determination before transforming the api/servers-call to a list of servers?

> > I wrote a quick small script that basically does what I need and grabs the first private_net and takes the IP from that if there is no external ip available. It seems to me that putting in some logic like that would make the plugin usable in both the cases described above.

This is the same approach I used for my own workaround, and I agree, this makes the plugin usable for my usecase.

> I feel uncomfortable "magically" deciding which way should be used to connect to the instance. Would you be happy if you could implement this yourself using [`compose`](https://docs.ansible.com/ansible/latest/plugins/inventory.html#id3) (related to #20)?

I would agree that this is magical, and in the case where there are several internal networks a machine belongs to, this would basically offer a random guess, which seems less than ideal.

I have to admit to not fully understanding the compose syntax and what it does, but according to the text surrounding the code it seems to do exactly what I want, so I would consider that an acceptable solution.

> > If that is not palatable, at least handling the fact that there are no primary ip by putting in the empty string seems to be a more robust way of handling a situation which would affect anyone who has at least one machine without a primary ipv4 address for their machine?

Would you consider a solution where the code basically creates a list of internal IPs, and if that list only contains one IP and the external IP is an empty string, putting the internal address there instead? Or does that break assumptions about the IP always being routable that I am not aware of?

To me, it seems that the desired priority would be to have the external IP first, and then the internal one if only one exists, and if there are multiple internal addresses, it would make sense to me to keep the ip address as the empty string, and add them as a list to a separate part of the yaml, so that it would be a single item list in the case of only one IP, and then I assume I would be able to use the compose-construction on that list to get the desired IP address. 

> For me so far, the only reason to use an inventory is because I want to connect to these instances and do something with them. I would rather be notified if any of the instances I selected (by using `label_selector` or other filter) could not be connected to in the way I intended (using `connect_with`).

I think that for me, it would make sense to have that logic on my end, either by assigning specific subnets to separate groups or doing what I am currently doing and assuming that the user has configured their own ssh client to their own desires.

I am not sure it's an advantage to make the assumption that one definition of "connectable" is all-encompassing. There seems to be several opportunities to use this data in a flexible way by just being an api to ansible-inventory-yaml-converter that makes a good effort to provide you with the best-guess value when the default value is not available. I don't believe the job of the inventory is to sanity check the connectability of the clients, as my (admittedly edge case) shows. Is there value to you in getting that notification when gathering the inventory, rather than when running your playbook?

> I think that the default for this is broken, and we should not have specified any default for the connection, and rather let the user decide. Either through setting `connect_with` or by setting `compose.ansible_host: SOMETHING`.

I am not sure what you mean with the default, but I think the above discussion hopefully leads us to a good solution.

> Based on your comments and my further research, I propose following changes:
> 
>     * Only set the `ansible_host` to what `connect_with` specifies if the value is actually available, possible log warning otherwise

As noted above, I would prefer to add a section with the internal IPs if there are only one and doing what seems to be a reasonable guess when there exists no external and only one internal IP. But I concede that your solution seems to solve the problem in an adequate way, and as long as the internal IP list is actually available in the data structure, I would be fine using compose.

>     * Fix the `compose`-ability to allow users to decide how they want to connect to instances using `jinja2` templates

I don't know if this requires extra effort separate from my problem, but if the functionality exists, it would make sense to me to add what is necessary for that functionality to work.

Thank you for a reasoned and thoughtful set of questions to clarify the thinking around this issue, and taking the time to look into a solution!

**Comment by apricote** [comment]
I submitted three PRs to improve upon this, please take a look and give feedback if you have the time:

- #186
- #187 
- #188

**Comment by apricote** [comment]
The mentioned PRs were released as `1.10.0`.

The inventory will now keep working if one of the nodes does not have a public ipv4 address, and you can configure fallbacks to your desire using `compose`.

I am going to close the issue, please feel free to ask any remaining questions here, or in a new issue.
